code/src/main/java/com/onlinecontacttracing/servlets/delete-data/OldDataDeleter.java

```diff
@@ -0,0 +1,71 @@
+package com.onlinecontacttracing.servlets;
+
+import static com.googlecode.objectify.ObjectifyService.ofy;
+import com.onlinecontacttracing.storage.NotificationBatch;
+import com.onlinecontacttracing.storage.NegativeUserPlace;
+import com.onlinecontacttracing.storage.NegativeUserLocation;
+import com.onlinecontacttracing.storage.PositiveUserPlaces;
+import com.onlinecontacttracing.storage.PositiveUserLocations;
+import com.onlinecontacttracing.storage.PositiveUserContacts;
+import com.onlinecontacttracing.storage.Constants;
+import com.google.common.collect.Iterables;
+import com.googlecode.objectify.Key;
+import java.util.logging.Logger;
+import java.time.Instant;
+
+/**
+* This class stores the information needed to query old data from a given class and delete it.
+*/
+public enum OldDataDeleter {
```

Comment:
A slightly better structure would be to have an OldDataDeleter that is an abstract class, and has 3 abstract methods (getClass, getQueryString, and getMaxAge).

Then you could have 6 concrete implementations of this Abstract Class, which each implement those three methods (what you're doing here).

Then in the servlet you would say `new NotificationBatchDeleter().delete()`. and you'd have `class NotificationBatchDeleter extends OldDataDeleter`.

Does that make sense?

Reply:
Makes sense but that sounds like a lot more code for the same functionality. Additionally it will be a lot of repeated code because the only difference between deletions is specifying the class, maximum age and query. Cynthia even commented that the current code could be abstracted even more since there is already repeated code (in the different servlets).

I will change the implementation if you want me to but I would rather work on other code (i.e. code not yet implemented).

Reply:
Happy to discuss this in detail, but inheritance is a better pattern here. The pattern you're using here is called static composition - you're encoding different types in static constants that you are calling enums. Why is inheritance better than static compositon (SC)?

* __Scalability__: If you need to add more per-type specificity, in SC need to add another argument (or even, a lambda function). In inheritance you'd just add another method, and you could support functions more gracefully.
* __Flexibility__: If you wanted at all custom behavior for any one of these behaviors, today you'd have to modify all of them (with a gross conditional). In an ineritance structure, you could just call super.fnName(), and then customize it.
* __Consistency__: Java is an object oriented programming language. We try to use its language features in a way that is consistent with the design principles of the language. This isn't an intended use case for enums.

I think Cynthia's point about code reuse on the servlet is a good one. But the way to solve that would be to add an abstract class with multiple concrete implementations, rather than having a statically composed set of servlets with a common set of functions.

Reply:
Also maybe I'm just reading into that last comment, but it sounds like you're a little bit frusturated here. While I can understand that, I want to push back a little and remind you about the distinction between engineering and code we talked about a while back.

In the future feel free to get early feedback on a code organization/strategy if you're unsure of it. This wasn't the right one in this case (which makes it a good learning opprotunity!) but if you're frusturated about iteration speed, the answer is earlier feedback, not worse code organization.

Reply:
Okay its done